ruby_programming/computer_science/project_data_structures_algorithms.md

```diff
@@ -15,34 +15,42 @@ You'll build a simple binary search tree in this assignment. In this lesson, our
 
-  6. Write `#inorder`, `#preorder`, and `#postorder` methods which accept a block. Each method should traverse the tree in their respective depth-first order and yield each node to the provided block. As a bonus, make the method return an array of values if no block is given.
+  7. Write `#inorder`, `#preorder`, and `#postorder` methods that returns an array of values. Each method should traverse the tree in their respective depth-first order.
 
-  7. Write a `#depth` method which accepts a node and returns the depth(number of levels) beneath the node.
+  8. Write a `#depth` method which accepts a node and returns the depth(number of levels) beneath the node.
 
-  8. Write a `#balanced?` method which checks if the tree is balanced. A balanced tree is one where the difference between heights of left subtree and right subtree is not more than 1.
+  9. Write a `#balanced?` method which checks if the tree is balanced. A balanced tree is one where the difference between heights of left subtree and right subtree is not more than 1.
 
-  9. Write a `#rebalance!` method which rebalances an unbalanced tree.
-  **Tip:** You'll want to create a level-order array of the tree before passing the array back into the `#build_tree` method.
+  10. Write a `#rebalance` method which rebalances an unbalanced tree. **Tip:** You'll want to create a level-order array of the tree before passing the array back into the `#build_tree` method.
 
-  10. Write a simple driver script that does the following:
+  11. Write a simple driver script that does the following:
 
     1. Create a binary search tree from an array of random numbers (`Array.new(15) { rand(1..100) }`)
     2. Confirm that the tree is balanced by calling `#balanced?`
     3. Print out all elements in level, pre, post, and in order
     4. try to unbalance the tree by adding several numbers > 100
     5. Confirm that the tree is unbalanced by calling `#balanced?`
-    6. Balance the tree by calling `#rebalance!`
+    6. Balance the tree by calling `#rebalance`
     7. Confirm that the tree is balanced by calling `#balanced?`
     8. Print out all elements in level, pre, post, and in order
 
-Pat yourself on the back! As a super-duper bonus, notice how all the depth-first methods share a similar signature and are basically just a re-arrangement of the same 3 lines... try dynamically declaring the three methods using metaprogamming techniques like `#define_method`.
+  **Tip:** If you would like to visualize your binary search tree, here is a `#pretty_print` method that a student wrote and shared on Discord:
+    ~~~ruby
+    def pretty_print(node = root, prefix="", is_left = true)
+      pretty_print(node.right, "#{prefix}#{is_left ? "│ " : " "}", false) if node.right 
+      puts "#{prefix}#{is_left ? "└── " : "┌── "}#{node.data.to_s}"
+      pretty_print(node.left, "#{prefix}#{is_left ? " " : "│ "}", true) if node.left
+    end
+    ~~~
+
+  ### Super-Duper Bonus
```

Comment:
what do you think about just removing this section entirely?

Reply:
This is one area that I do not feel like I have enough Ruby experience to make a decision like this. I just heard that it was not a good use of meta-programming, but then it was suggested to stay in (with a disclaimer). https://github.com/TheOdinProject/curriculum/issues/17956
Maybe @leosoaivan can offer some valuable insight on this

Reply:
I mean removing the super-duper bonus

Reply:
I agree with you and will remove it. Thank you for your feedback!